File: cpp/compiler/tac_visitor.cpp
#include "tac_visitor.hpp"
#include "exceptions.hpp"

TacVisitor::TacVisitor(FunctionBody& body) : body(body), result(nullptr) {
}

void TacVisitor::visitProgram(ProgramNode* const node) {
	throw std::logic_error("ProgramNode should not be visited by TacVisitor");
}

void TacVisitor::visitFunctionDefinition(FunctionDefinitionNode* const node) {
    body.emplaceInstruction<FunctionInstruction>(node->lineNumber, body.name);
    body.emplaceInstruction<AllocateStackInstruction>(node->lineNumber);

    if (node->body) {
    	node->body->accept(*this);
    }
}

void TacVisitor::visitDeclaration(DeclarationNode* const node) {
    auto pseudoRegister = std::make_shared<PseudoRegister>(PseudoRegister{ *body.name, body.variableCount });
    body.variableToPseudoregister[*node->identifier] = pseudoRegister;
    if (node->expression) {
        node->expression->accept(*this);
        body.emplaceInstructionWithDestination<StoreValueInstruction>(node->lineNumber, pseudoRegister, result);
    }
    body.variableCount++;
}

void TacVisitor::visitAssignment(AssignmentNode* const node) {
    node->right->accept(*this);
    std::shared_ptr<Operand> src = result;
    try {
        node->left->accept(*this);
        auto variable = std::get<std::shared_ptr<PseudoRegister>>(*result);
        //node->right->accept(*this);
        //Operand src = result;
        body.emplaceInstructionWithDestination<StoreValueInstruction>(node->lineNumber, variable, src);
    } catch (std::bad_variant_access&) {
        throw semantic_error(std::format("Invalid lvalue {} at {}", result, node->lineNumber));
    }
}

void TacVisitor::visitReturn(ReturnNode* const node) {
    std::shared_ptr<Operand> dest = nullptr;
    if (node->expression) {
        node->expression->accept(*this);
        dest = result;
    }
    body.emplaceInstruction<ReturnInstruction>(node->lineNumber, dest);
    result = nullptr;
}

void TacVisitor::visitUnary(UnaryNode* const node) {
    node->expression->accept(*this);
    if (node->op == UnaryOperator::UNARY_ADD) { // if unary add do nothing
        return;
    }
    std::shared_ptr<Operand> src = result;
    auto dest = body.emplaceInstruction<UnaryOpInstruction>(node->lineNumber, node->op, src);
    body.variableCount++;
    result = std::make_shared<Operand>(dest);
}

void TacVisitor::visitBinary(BinaryNode* const node) {
    if (node->op == BinaryOperator::LOGICAL_AND) {
        auto falseLabel = std::make_shared<std::string>(std::format(".{}{}_false", body.name, ++body.labelCount));
        auto endLabel = std::make_shared<std::string>(std::format(".{}{}_end", body.name, ++body.labelCount));

        // Short-circuiting
        node->left->accept(*this);
        auto leftOperand = result;
        body.emplaceInstruction<JumpIfZero>(node->lineNumber, leftOperand, falseLabel); // goto false label

        node->right->accept(*this);
        auto rightOperand = result;
        body.emplaceInstruction<JumpIfZero>(node->lineNumber, rightOperand, falseLabel);

        auto dest = body.emplaceInstruction<StoreValueInstruction>(node->lineNumber, std::make_shared<Operand>(static_cast<Number>(1)));
        body.emplaceInstruction<Jump>(node->lineNumber, endLabel); // goto end

        body.emplaceInstruction<Label>(node->lineNumber, falseLabel); // false label
        dest = body.emplaceInstruction<StoreValueInstruction>(node->lineNumber, std::make_shared<Operand>(static_cast<Number>(0)));

        body.emplaceInstruction<Label>(node->lineNumber, endLabel); // end
        body.variableCount++;
        result = std::make_shared<Operand>(dest);
        return;
    }

    if (node->op == BinaryOperator::LOGICAL_OR) {
        auto trueLabel = std::make_shared<std::string>(std::format(".{}{}_true", body.name, ++body.labelCount));
        auto endLabel = std::make_shared<std::string>(std::format(".{}{}_end", body.name, ++body.labelCount));

        // Short-circuiting
        node->left->accept(*this);
        auto leftOperand = result;
        body.emplaceInstruction<JumpIfNotZero>(node->lineNumber, leftOperand, trueLabel); // goto true label

        node->right->accept(*this);
        auto rightOperand = result;
        body.emplaceInstruction<JumpIfNotZero>(node->lineNumber, rightOperand, trueLabel);

        auto dest = body.emplaceInstruction<StoreValueInstruction>(node->lineNumber, std::make_shared<Operand>(static_cast<Number>(0)));
        body.emplaceInstruction<Jump>(node->lineNumber, endLabel); // goto end

        body.emplaceInstruction<Label>(node->lineNumber, trueLabel); // true label
        dest = body.emplaceInstruction<StoreValueInstruction>(node->lineNumber, std::make_shared<Operand>(static_cast<Number>(1)));

        body.emplaceInstruction<Label>(node->lineNumber, endLabel); // end
        body.variableCount++;
        result = std::make_shared<Operand>(dest);
        return;
    }

    node->left->accept(*this);
    auto leftOperand = result;

    node->right->accept(*this);
    auto rightOperand = result;

    auto dest = body.emplaceInstruction<BinaryOpInstruction>(node->lineNumber, node->op, leftOperand, rightOperand);
    body.variableCount++;
    result = std::make_shared<Operand>(dest);
}

void TacVisitor::visitConst(ConstNode* const node) {
    result = std::make_shared<Operand>(node->value);
}

void TacVisitor::visitVariable(VariableNode* const node) {
    if (!body.variableToPseudoregister.contains(*node->identifier)) {
        throw semantic_error(std::format("Undeclared variable {} at {}", node->identifier, node->lineNumber));
    }
    result = std::make_shared<Operand>(body.variableToPseudoregister[*node->identifier]);
}

void TacVisitor::visitPostfix(PostfixNode* const node) {
    node->variable->accept(*this); // get variable
    auto variable = result; // save variable
    auto temp1 = body.emplaceInstruction<StoreValueInstruction>(node->lineNumber, result); // temp1 = a
    body.variableCount++;
    auto temp2 = body.emplaceInstruction<BinaryOpInstruction>(node->lineNumber, node->op,
        variable, std::make_shared<Operand>(static_cast<Number>(1))); // t2 = a + 1
    ++body.variableCount;
    body.emplaceInstructionWithDestination<StoreValueInstruction>(node->lineNumber, std::get<std::shared_ptr<PseudoRegister>>(*variable), std::make_shared<Operand>(temp2)); // a = t2
    result = std::make_shared<Operand>(temp1);
}

void TacVisitor::visitPrefix(PrefixNode* const node) {
    node->variable->accept(*this);
    auto variable = result;
    body.emplaceInstructionWithDestination<BinaryOpInstruction>(node->lineNumber, std::get<std::shared_ptr<PseudoRegister>>(*variable), node->op, variable, std::make_shared<Operand>(static_cast<Number>(1)));
    body.variableCount++;
}

/*
 <instructions for condition>
c = <result of condition>
JumpIfZero(c, end)
<instructions for statement>
Label(end)
*/
void TacVisitor::visitCondition(ConditionNode* const node) {
    if (node->isTernary) {
        node->condition->accept(*this);
        auto condition = result;
        auto elseLabel = std::make_shared<std::string>(std::format(".{}{}_else", body.name, ++body.labelCount));
        auto endLabel = std::make_shared<std::string>(std::format(".{}{}_end", body.name, ++body.labelCount));
        auto dest = std::make_shared<PseudoRegister>(PseudoRegister{ *body.name, body.variableCount++ });
        body.emplaceInstruction<JumpIfZero>(node->lineNumber, condition, elseLabel); // if false goto else
        node->ifTrue->accept(*this);
        body.emplaceInstructionWithDestination<StoreValueInstruction>(node->lineNumber, dest, result);
        body.emplaceInstruction<Jump>(node->lineNumber, endLabel); // goto end
        body.emplaceInstruction<Label>(node->lineNumber, elseLabel); // else
        node->ifFalse->accept(*this);
        body.emplaceInstructionWithDestination<StoreValueInstruction>(node->lineNumber, dest, result);
        body.emplaceInstruction<Label>(node->lineNumber, endLabel); // end
        result = std::make_shared<Operand>(dest);
    } else if (node->ifFalse == nullptr) {
		node->condition->accept(*this);
		auto condition = result;
		auto endLabel = std::make_shared<std::string>(std::format(".{}{}_end", body.name, ++body.labelCount));
		body.emplaceInstruction<JumpIfZero>(node->lineNumber, condition, endLabel); // if false goto end
		node->ifTrue->accept(*this);
		body.emplaceInstruction<Label>(node->lineNumber, endLabel); // end
        result = nullptr;
    } else {
		node->condition->accept(*this);
		auto condition = result;
		auto elseLabel = std::make_shared<std::string>(std::format(".{}{}_else", body.name, ++body.labelCount));
		auto endLabel = std::make_shared<std::string>(std::format(".{}{}_end", body.name, ++body.labelCount));
		body.emplaceInstruction<JumpIfZero>(node->lineNumber, condition, elseLabel); // if false goto else
		node->ifTrue->accept(*this);
		body.emplaceInstruction<Jump>(node->lineNumber, endLabel); // goto end
		body.emplaceInstruction<Label>(node->lineNumber, elseLabel); // else
		node->ifFalse->accept(*this);
		body.emplaceInstruction<Label>(node->lineNumber, endLabel); // end
        result = nullptr;
    }
}

void TacVisitor::visitBlock(BlockNode* const node) {
	for (auto& item : node->block_items) {
		item->accept(*this);
	}
}

/*
Label(start)
<instructions for condition>
v = <result of condition>
JumpIfZero(v, end)
<instructions for body>
1 Jump(start)
Label(end)
 */
void TacVisitor::visitWhile(WhileNode* const node) {
	auto startLabel = std::make_shared<std::string>(std::format(".{}{}_start.loop", body.name, node->label));
	auto endLabel = std::make_shared<std::string>(std::format(".{}{}_end.loop", body.name, node->label));
	body.emplaceInstruction<Label>(node->lineNumber, startLabel); // start
	node->condition->accept(*this);
	auto condition = result;
	body.emplaceInstruction<JumpIfZero>(node->lineNumber, condition, endLabel); // if false goto end
    if (node->body) {
        node->body->accept(*this);
    }
	body.emplaceInstruction<Jump>(node->lineNumber, startLabel); // goto start
	body.emplaceInstruction<Label>(node->lineNumber, endLabel); // end
	result = nullptr;
}

void TacVisitor::visitBreak(BreakNode* const node) {
	body.emplaceInstruction<Jump>(node->lineNumber, std::make_shared<std::string>(std::format(".{}{}_end.loop", body.name, node->label)));
}

void TacVisitor::visitContinue(ContinueNode* const node) {
	if (node->isFor) {
		body.emplaceInstruction<Jump>(node->lineNumber, std::make_shared<std::string>(std::format(".{}{}_increment.loop", body.name, node->label)));
	} else {
		body.emplaceInstruction<Jump>(node->lineNumber, std::make_shared<std::string>(std::format(".{}{}_start.loop", body.name, node->label)));
	}
}

void TacVisitor::visitFor(ForNode* const node) {
	auto startLabel = std::make_shared<std::string>(std::format(".{}{}_start.loop", body.name, node->label));
	auto endLabel = std::make_shared<std::string>(std::format(".{}{}_end.loop", body.name, node->label));
	auto incrementLabel = std::make_shared<std::string>(std::format(".{}{}_increment.loop", body.name, node->label));
	if (node->init) {
		node->init->accept(*this);
	}
	body.emplaceInstruction<Label>(node->lineNumber, startLabel); // start
	if (node->condition) {
		node->condition->accept(*this);
		auto condition = result;
		body.emplaceInstruction<JumpIfZero>(node->lineNumber, condition, endLabel); // if false goto end
	}
	if (node->body) {
		node->body->accept(*this);
	}
    body.emplaceInstruction<Label>(node->lineNumber, incrementLabel); // increment
	if (node->increment) {
		node->increment->accept(*this);
	}
	body.emplaceInstruction<Jump>(node->lineNumber, startLabel); // goto start
	body.emplaceInstruction<Label>(node->lineNumber, endLabel); // end
	result = nullptr;
}

File: rust/src/tac_visitor.rs
use crate::ast::{ASTNode, Visitor};
use crate::common::{Operand, Position, Pseudoregister};
use crate::errors::CompilerError;
use crate::errors::CompilerError::SemanticError;
use crate::lexer::{BinaryOperator, Number, UnaryOperator};
use crate::tac::FunctionBody;
use crate::tac::TACInstructionType::{
    BinaryOpInstruction, FunctionInstruction, Jump, JumpIfZero, Label, ReturnInstruction,
    StoreValueInstruction, UnaryOpInstruction,
};
use std::rc::Rc;

pub struct TacVisitor<'a> {
    name: Rc<String>,
    body: &'a mut FunctionBody,
    result: Rc<Operand>,
}

impl<'a> TacVisitor<'a> {
    pub fn new(name: Rc<String>, body: &'a mut FunctionBody) -> Self {
        Self {
            name,
            body,
            result: Rc::new(Operand::None),
        }
    }
}

impl<'a> Visitor for TacVisitor<'a> {
    fn visit_program(
        &mut self,
        _line_number: &Rc<Position>,
        _function_declaration: &mut Box<ASTNode>,
    ) -> Result<(), CompilerError> {
        panic!("Should not be called")
    }

    fn visit_function(
        &mut self,
        line_number: &Rc<Position>,
        identifier: &mut Rc<String>,
        body: &mut Option<Box<ASTNode>>,
    ) -> Result<(), CompilerError> {
        self.body.add_instruction(
            Rc::clone(&line_number),
            FunctionInstruction {
                name: identifier.clone(),
            },
        );
        if let Some(body) = body {
            body.accept(self)
        } else {
            Ok(())
        }
    }

    fn visit_declaration(
        &mut self,
        line_number: &Rc<Position>,
        identifier: &mut Rc<String>,
        expression: &mut Option<Box<ASTNode>>,
    ) -> Result<(), CompilerError> {
        let pseudoregister = Rc::new(Pseudoregister::new(
            (*self.name).clone(),
            self.body.variable_count,
        ));
        self.body
            .variable_to_pseudoregister
            .insert(identifier.clone(), Rc::clone(&pseudoregister));
        if let Some(expression) = expression {
            expression.accept(self)?;
            self.body.add_instruction(
                line_number.clone(),
                StoreValueInstruction {
                    dest: Rc::clone(&pseudoregister),
                    src: Rc::clone(&self.result),
                },
            );
        }
        self.body.variable_count += 1;
        Ok(())
    }

    fn visit_assignment(
        &mut self,
        line_number: &Rc<Position>,
        left: &mut Box<ASTNode>,
        right: &mut Box<ASTNode>,
    ) -> Result<(), CompilerError> {
        right.accept(self)?;
        let src = Rc::clone(&self.result);
        left.accept(self)?;
        let dest = Rc::clone(&self.result);
        match dest.as_ref() {
            Operand::Register(variable) => {
                let dest_registry: Rc<Pseudoregister> = Rc::clone(variable);
                self.body.add_instruction(
                    line_number.clone(),
                    StoreValueInstruction {
                        dest: dest_registry,
                        src,
                    },
                );
                Ok(())
            }
            _ => Err(SemanticError(format!(
                "Expected lvalue but got {:?} at {:?}",
                src, line_number
            ))),
        }
    }

    fn visit_return(
        &mut self,
        line_number: &Rc<Position>,
        expression: &mut Option<Box<ASTNode>>,
    ) -> Result<(), CompilerError> {
        if let Some(expression) = expression {
            expression.accept(self)?;
            self.body.add_instruction(
                line_number.clone(),
                ReturnInstruction {
                    val: Rc::clone(&self.result),
                },
            );
        } else {
            self.body.add_instruction(
                line_number.clone(),
                ReturnInstruction {
                    val: Rc::new(Operand::None),
                },
            );
        }
        Ok(())
    }

    fn visit_block(
        &mut self,
        _line_number: &Rc<Position>,
        body: &mut Vec<Box<ASTNode>>,
    ) -> Result<(), CompilerError> {
        for item in body {
            item.accept(self)?;
        }
        Ok(())
    }

    fn visit_unary(
        &mut self,
        line_number: &Rc<Position>,
        op: &mut UnaryOperator,
        expression: &mut Box<ASTNode>,
    ) -> Result<(), CompilerError> {
        expression.accept(self)?;
        if *op == UnaryOperator::UnaryAdd {
            return Ok(());
        }
        let src = Rc::clone(&self.result);
        let dest = Rc::new(Pseudoregister::new(
            (*self.name).clone(),
            self.body.variable_count,
        ));
        self.body.variable_count += 1;
        self.body.add_instruction(
            line_number.clone(),
            UnaryOpInstruction {
                dest: Rc::clone(&dest),
                op: *op,
                operand: src,
            },
        );
        self.result = Rc::from(Operand::Register(Rc::clone(&dest)));
        Ok(())
    }

    fn visit_binary(
        &mut self,
        line_number: &Rc<Position>,
        op: &mut BinaryOperator,
        left: &mut Box<ASTNode>,
        right: &mut Box<ASTNode>,
    ) -> Result<(), CompilerError> {
        match op {
            BinaryOperator::LogicalAnd => {
                let false_label: Rc<String> =
                    Rc::from(format!("{}{}_false", self.name, self.body.label_count));
                self.body.label_count += 1;
                let end_label: Rc<String> =
                    Rc::from(format!("{}{}_end", self.name, self.body.label_count));
                self.body.label_count += 1;

                // Short circuiting
                left.accept(self)?;
                let left_operand = Rc::clone(&self.result);
                self.body.add_instruction(
                    line_number.clone(),
                    JumpIfZero {
                        label: false_label.clone(),
                        operand: left_operand,
                    },
                ); // goto false

                right.accept(self)?;
                let right_operand = Rc::clone(&self.result);
                self.body.add_instruction(
                    line_number.clone(),
                    JumpIfZero {
                        label: false_label.clone(),
                        operand: right_operand,
                    },
                ); // goto false

                let dest = Rc::new(Pseudoregister::new(
                    (*self.name).clone(),
                    self.body.variable_count,
                ));
                self.body.add_instruction(
                    line_number.clone(),
                    StoreValueInstruction {
                        dest: Rc::clone(&dest),
                        src: Rc::new(Operand::Immediate(1)),
                    },
                );
                self.body.add_instruction(
                    line_number.clone(),
                    Jump {
                        label: end_label.clone(),
                    },
                ); // goto end

                // false label
                self.body.add_instruction(
                    line_number.clone(),
                    Label {
                        label: false_label.clone(),
                    },
                );
                self.body.add_instruction(
                    line_number.clone(),
                    StoreValueInstruction {
                        dest: Rc::clone(&dest),
                        src: Rc::new(Operand::Immediate(0)),
                    },
                );

                // end
                self.body.add_instruction(
                    line_number.clone(),
                    Label {
                        label: end_label.clone(),
                    },
                );
                self.result = Rc::from(Operand::Register(dest));
                Ok(())
            }
            BinaryOperator::LogicalOr => todo!(),
            _ => {
                left.accept(self)?;
                let left = Rc::clone(&self.result);

                right.accept(self)?;
                let right = Rc::clone(&self.result);

                let dest = Rc::new(Pseudoregister::new(
                    (*self.name).clone(),
                    self.body.variable_count,
                ));
                self.body.variable_count += 1;
                self.body.add_instruction(
                    line_number.clone(),
                    BinaryOpInstruction {
                        dest: Rc::clone(&dest),
                        op: *op,
                        left,
                        right,
                    },
                );
                self.result = Rc::from(Operand::Register(dest));
                Ok(())
            }
        }
    }

    fn visit_condition(
        &mut self,
        line_number: &Rc<Position>,
        condition: &mut Box<ASTNode>,
        if_true: &mut Option<Box<ASTNode>>,
        if_false: &mut Option<Box<ASTNode>>,
    ) -> Result<(), CompilerError> {
        todo!()
    }

    fn visit_while(
        &mut self,
        line_number: &Rc<Position>,
        condition: &mut Box<ASTNode>,
        body: &mut Option<Box<ASTNode>>,
        label: &mut Rc<String>,
        is_do_while: &mut bool,
    ) -> Result<(), CompilerError> {
        todo!()
    }

    fn visit_break(
        &mut self,
        line_number: &Rc<Position>,
        label: &mut Rc<String>,
    ) -> Result<(), CompilerError> {
        todo!()
    }

    fn visit_continue(
        &mut self,
        line_number: &Rc<Position>,
        label: &mut Rc<String>,
        is_for: &mut bool,
    ) -> Result<(), CompilerError> {
        todo!()
    }

    fn visit_for(
        &mut self,
        line_number: &Rc<Position>,
        init: &mut Option<Box<ASTNode>>,
        condition: &mut Option<Box<ASTNode>>,
        increment: &mut Option<Box<ASTNode>>,
        body: &mut Option<Box<ASTNode>>,
        label: &mut Rc<String>,
    ) -> Result<(), CompilerError> {
        todo!()
    }

    fn visit_const(
        &mut self,
        line_number: &Rc<Position>,
        value: &mut Number,
    ) -> Result<(), CompilerError> {
        todo!()
    }

    fn visit_variable(
        &mut self,
        line_number: &Rc<Position>,
        identifier: &mut Rc<String>,
    ) -> Result<(), CompilerError> {
        todo!()
    }

    fn visit_prefix(
        &mut self,
        line_number: &Rc<Position>,
        variable: &mut Box<ASTNode>,
        operator: &mut UnaryOperator,
    ) -> Result<(), CompilerError> {
        todo!()
    }

    fn visit_postfix(
        &mut self,
        line_number: &Rc<Position>,
        variable: &mut Box<ASTNode>,
        operator: &mut UnaryOperator,
    ) -> Result<(), CompilerError> {
        todo!()
    }
}
